src/Configuration/Programmer/Programmer_helper.cpp

```diff
@@ -0,0 +1,453 @@
+
+std::string transportTypeToString(TransportType transport) {
+  switch (transport) {
+    case TransportType::jtag:
+      return "jtag";
+      // Handle other transport types as needed
+  }
+  return "";
+}
+
+int get_string_descriptor(struct libusb_device_handle* device,
+                          uint8_t desc_index, std::string& stringDesc,
+                          std::string& outputMsg) {
+  int retval = 0;
+  char desc_string[256]; /* Max size of string descriptor */
+  retval = libusb_get_string_descriptor_ascii(
+      device, desc_index, (unsigned char*)desc_string, sizeof(desc_string));
+  stringDesc = desc_string;
+  if (retval < 0) {
+    outputMsg += "libusb_get_string_descriptor_ascii() failed with " +
+                 std::string(libusb_error_name(retval)) + "\n";
+    return retval;
+  }
+  return retval;
+}
+
+void addOrUpdateErrorMessage(int error, const std::string& message) {
+  auto it = ErrorMessages.find(error);
+  if (it != ErrorMessages.end()) {
+    // Error code already exists, update the error message
+    it->second = message;
+  } else {
+    // Error code doesn't exist, insert the new error code and message
+    ErrorMessages.insert({error, message});
+  }
+}
+
+std::vector<std::string> findStringPattern(const std::string& input,
+                                           const std::string& pattern) {
+  std::vector<std::string> matches;
+  std::regex re(pattern);
+  std::sregex_iterator next(input.begin(), input.end(), re);
+  std::sregex_iterator end;
+  while (next != end) {
+    std::smatch match = *next;
+    matches.push_back(match.str());
+    next++;
+  }
+  return matches;
+}
+
+int isCableSupported(const Cable& cable) {
+  bool cableSupported = false;
+  for (auto& supportedCable : supportedCableVendorIdProductId) {
+    if (cable.vendorId == std::get<0>(supportedCable) &&
+        cable.productId == std::get<1>(supportedCable)) {
+      cableSupported = true;
+      break;
+    }
+  }
+  if (!cableSupported) {
+    return ProgrammerErrorCode::CableNotSupported;
+  }
+  return ProgrammerErrorCode::NoError;
+}
+
+/*
+This function expecting the string from openOCD scan_chain output
+with example below
+
+testString =
+"   TapName            Enabled IdCode     Expected   IrLen IrCap IrMask\n"
+"-- ------------------ ------- ---------- ---------- ----- ----- ------\n"
+"0  omap5912.dsp          Y    0x03df1d81 0x03df1d81 38    0x01  0x03\n"
+"1  omap5912.arm          Y    0x0692602f 0x0692602f 4     0x01  0x0f\n"
+"2  omap5912.unknown      Y    0x00000000 0x00000000 8     0x01  0x03\n"
+"3 auto0.tap              Y     0x20000913 0x00000000     5 0x01  0x03
+^\d+\s+\w+\.\w+\s+\w+\s+0x[\da-fA-F]+\s+0x[\da-fA-F]+\s+\d+\s+0x[\da-fA-F]+\s+0x[\da-fA-F]+$
+*/
+std::vector<TapInfo> extractTapInfoList(const std::string& tapInfoString) {
+  std::vector<TapInfo> tapInfoList;
+  std::istringstream iss(tapInfoString);
+  std::string line;
+  std::string token;
+  size_t pos = -1;
```

Comment:
size_t is unsigned type. Looks suspicious

Reply:
Yes. I did this with purpose to initialize the pos variable to -1. It will end up as a very large unsigned data as a invalid number. A zero may give some meaningful value.

Reply:
GY, you should change this to std::string::npos. They are equivalent, but using npos is more obvious -- to initiailize with invalid pos.

Reply:
Noted. Make sense.

Reply:
Fixed. Change to std::string:npos